File: src/com/maddyhome/idea/vim/group/FileGroup.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2019 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.group;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.LogicalPosition;
import com.intellij.openapi.fileEditor.*;
import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx;
import com.intellij.openapi.fileEditor.impl.EditorWindow;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.openapi.fileTypes.FileTypeManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.util.Ref;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VfsUtilCore;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileVisitor;
import com.maddyhome.idea.vim.KeyHandler;
import com.maddyhome.idea.vim.VimPlugin;
import com.maddyhome.idea.vim.command.CommandState;
import com.maddyhome.idea.vim.common.TextRange;
import com.maddyhome.idea.vim.helper.EditorHelper;
import com.maddyhome.idea.vim.helper.SearchHelper;
import com.maddyhome.idea.vim.helper.StringHelper;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.io.File;
import java.util.HashMap;

public class FileGroup {
  public boolean openFile(@NotNull String filename, @NotNull DataContext context) {
    if (logger.isDebugEnabled()) {
      logger.debug("openFile(" + filename + ")");
    }
    final Project project = PlatformDataKeys.PROJECT.getData(context); // API change - don't merge
    if (project == null) return false;

    VirtualFile found = findFile(filename, project);

    if (found != null) {
      if (logger.isDebugEnabled()) {
        logger.debug("found file: " + found);
      }
      // Can't open a file unless it has a known file type. The next call will return the known type.
      // If unknown, IDEA will prompt the user to pick a type.
      FileType type = FileTypeManager.getInstance().getKnownFileTypeOrAssociate(found, project);
      if (type != null) {
        FileEditorManager fem = FileEditorManager.getInstance(project);
        fem.openFile(found, true);

        return true;
      }
      else {
        // There was no type and user didn't pick one. Don't open the file
        // Return true here because we found the file but the user canceled by not picking a type.
        return true;
      }
    }
    else {
      VimPlugin.showMessage("Unable to find " + filename);

      return false;
    }
  }

  @Nullable
  VirtualFile findFile(@NotNull String filename, @NotNull Project project) {
    VirtualFile found = null;
    if (filename.length() > 2 && filename.charAt(0) == '~' && filename.charAt(1) == File.separatorChar) {
      String homefile = filename.substring(2);
      String dir = System.getProperty("user.home");
      if (logger.isDebugEnabled()) {
        logger.debug("home dir file");
        logger.debug("looking for " + homefile + " in " + dir);
      }
      found = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(new File(dir, homefile));
    }
    else {
      ProjectRootManager prm = ProjectRootManager.getInstance(project);
      VirtualFile[] roots = prm.getContentRoots();
      for (int i = 0; i < roots.length; i++) {
        if (logger.isDebugEnabled()) {
          logger.debug("root[" + i + "] = " + roots[i].getPath());
        }
        found = findFile(roots[i], filename);
        if (found != null) {
          break;
        }
      }

      if (found == null) {
        found = LocalFileSystem.getInstance().findFileByIoFile(new File(filename));
      }
    }

    return found;
  }

  @Nullable
  private VirtualFile findFile(@NotNull VirtualFile root, @NotNull String filename) {
    VirtualFile res = root.findFileByRelativePath(filename);
    if (res != null) {
      return res;
    }
    final Ref<VirtualFile> result = Ref.create();
    final VirtualFileVisitor<Object> visitor = new VirtualFileVisitor<Object>() {
      @Override
      public boolean visitFile(@NotNull VirtualFile file) {
        if (file.getName().equals(filename)) {
          result.set(file);
          return false;
        }
        return true;
      }
    };
    VfsUtilCore.visitChildrenRecursively(root, visitor);
    return result.get();
  }

  /**
   * Closes the current editor.
   */
  public void closeFile(@NotNull Editor editor, @NotNull DataContext context) {
    final Project project = PlatformDataKeys.PROJECT.getData(context);
    if (project != null) {
      final FileEditorManagerEx fileEditorManager = FileEditorManagerEx.getInstanceEx(project);
      final EditorWindow window = fileEditorManager.getCurrentWindow();
      final VirtualFile virtualFile = EditorHelper.getVirtualFile(editor);

      if (virtualFile != null && window != null) {
        window.closeFile(virtualFile);
      }
    }
  }

  /**
   * Saves specific file in the project.
   */
  public void saveFile(DataContext context) {
    KeyHandler.executeAction("SaveDocument", context);
  }

  /**
   * Saves all files in the project.
   */
  public void saveFiles(DataContext context) {
    KeyHandler.executeAction("SaveAll", context);
  }

  /**
   * Selects then next or previous editor.
   */
  public boolean selectFile(int count, @NotNull DataContext context) {
    final Project project = PlatformDataKeys.PROJECT.getData(context);
    if (project == null) return false;
    FileEditorManager fem = FileEditorManager.getInstance(project); // API change - don't merge
    VirtualFile[] editors = fem.getOpenFiles();
    if (count == 99) {
      count = editors.length - 1;
    }
    if (count < 0 || count >= editors.length) {
      return false;
    }

    fem.openFile(editors[count], true);

    return true;
  }

  /**
   * Selects then next or previous editor.
   */
  public void selectNextFile(int count, @NotNull DataContext context) {
    Project project = PlatformDataKeys.PROJECT.getData(context);
    if (project == null) return;
    FileEditorManager fem = FileEditorManager.getInstance(project); // API change - don't merge
    VirtualFile[] editors = fem.getOpenFiles();
    VirtualFile current = fem.getSelectedFiles()[0];
    for (int i = 0; i < editors.length; i++) {
      if (editors[i].equals(current)) {
        int pos = (i + (count % editors.length) + editors.length) % editors.length;

        fem.openFile(editors[pos], true);
      }
    }
  }

  /**
   * Selects previous editor tab.
   */
  public void selectPreviousTab(@NotNull DataContext context) {
    Project project = PlatformDataKeys.PROJECT.getData(context);
    if (project == null) return;
    FileEditorManager fem = FileEditorManager.getInstance(project); // API change - don't merge
    VirtualFile vf = lastSelections.get(fem);
    if (vf != null && vf.isValid()) {
      fem.openFile(vf, true);
    }
    else {
      VimPlugin.indicateError();
    }
  }

  @Nullable
  Editor selectEditor(Project project, @NotNull VirtualFile file) {
    FileEditorManager fMgr = FileEditorManager.getInstance(project);
    FileEditor[] feditors = fMgr.openFile(file, true);
    if (feditors.length > 0) {
      if (feditors[0] instanceof TextEditor) {
        Editor editor = ((TextEditor) feditors[0]).getEditor();
        if (!editor.isDisposed()) {
          return editor;
        }
      }
    }

    return null;
  }

  public void displayAsciiInfo(@NotNull Editor editor) {
    int offset = editor.getCaretModel().getOffset();
    CharSequence charsSequence = editor.getDocument().getCharsSequence();
    if (charsSequence.length() == 0 || offset >= charsSequence.length()) return;
    char ch = charsSequence.charAt(offset);

    VimPlugin.showMessage("<" +
                          StringHelper.toKeyNotation(KeyStroke.getKeyStroke(ch)) +
                          ">  " +
                          (int)ch +
                          ",  Hex " +
                          Long.toHexString(ch) +
                          ",  Octal " +
                          Long.toOctalString(ch));
  }

  public void displayHexInfo(@NotNull Editor editor) {
    int offset = editor.getCaretModel().getOffset();
    char ch = editor.getDocument().getCharsSequence().charAt(offset);

    VimPlugin.showMessage(Long.toHexString(ch));
  }

  public void displayLocationInfo(@NotNull Editor editor) {
    StringBuilder msg = new StringBuilder();
    Document doc = editor.getDocument();

    if (CommandState.getInstance(editor).getMode() != CommandState.Mode.VISUAL) {
      LogicalPosition lp = editor.getCaretModel().getLogicalPosition();
      int col = editor.getCaretModel().getOffset() - doc.getLineStartOffset(lp.line);
      int endoff = doc.getLineEndOffset(lp.line);
      if (doc.getCharsSequence().charAt(endoff) == '\n') {
        endoff--;
      }
      int ecol = endoff - doc.getLineStartOffset(lp.line);
      LogicalPosition elp = editor.offsetToLogicalPosition(endoff);

      msg.append("Col ").append(col + 1);
      if (col != lp.column) {
        msg.append("-").append(lp.column + 1);
      }

      msg.append(" of ").append(ecol + 1);
      if (ecol != elp.column) {
        msg.append("-").append(elp.column + 1);
      }

      int lline = editor.getCaretModel().getLogicalPosition().line;
      int total = EditorHelper.getLineCount(editor);

      msg.append("; Line ").append(lline + 1).append(" of ").append(total);

      SearchHelper.CountPosition cp = SearchHelper.countWords(editor);

      msg.append("; Word ").append(cp.getPosition()).append(" of ").append(cp.getCount());

      int offset = editor.getCaretModel().getOffset();
      int size = EditorHelper.getFileSize(editor);

      msg.append("; Character ").append(offset + 1).append(" of ").append(size);
    }
    else {
      msg.append("Selected ");

      TextRange vr = new TextRange(editor.getSelectionModel().getBlockSelectionStarts(), editor.getSelectionModel().getBlockSelectionEnds());
      vr.normalize();

      int lines;
      SearchHelper.CountPosition cp = SearchHelper.countWords(editor);
      int words = cp.getCount();
      int word = 0;
      if (vr.isMultiple()) {
        lines = vr.size();
        int cols = vr.getMaxLength();

        msg.append(cols).append(" Cols; ");

        for (int i = 0; i < vr.size(); i++) {
          cp = SearchHelper.countWords(editor, vr.getStartOffsets()[i], vr.getEndOffsets()[i] - 1);
          word += cp.getCount();
        }
      }
      else {
        LogicalPosition slp = editor.offsetToLogicalPosition(vr.getStartOffset());
        LogicalPosition elp = editor.offsetToLogicalPosition(vr.getEndOffset());

        lines = elp.line - slp.line + 1;

        cp = SearchHelper.countWords(editor, vr.getStartOffset(), vr.getEndOffset() - 1);
        word = cp.getCount();
      }

      int total = EditorHelper.getLineCount(editor);

      msg.append(lines).append(" of ").append(total).append(" Lines");

      msg.append("; ").append(word).append(" of ").append(words).append(" Words");

      int chars = vr.getSelectionCount();
      int size = EditorHelper.getFileSize(editor);

      msg.append("; ").append(chars).append(" of ").append(size).append(" Characters");
    }

    VimPlugin.showMessage(msg.toString());
  }

  public void displayFileInfo(@NotNull Editor editor, boolean fullPath) {
    StringBuilder msg = new StringBuilder();
    VirtualFile vf = EditorHelper.getVirtualFile(editor);
    if (vf != null) {
      msg.append('"');
      if (fullPath) {
        msg.append(vf.getPath());
      }
      else {
        Project project = editor.getProject();
        if (project != null) {
          VirtualFile root = ProjectRootManager.getInstance(project).getFileIndex().getContentRootForFile(vf);
          if (root != null) {
            msg.append(vf.getPath().substring(root.getPath().length() + 1));
          }
          else {
            msg.append(vf.getPath());
          }
        }
      }
      msg.append("\" ");
    }
    else {
      msg.append("\"[No File]\" ");
    }

    Document doc = editor.getDocument();
    if (!doc.isWritable()) {
      msg.append("[RO] ");
    }
    else if (FileDocumentManager.getInstance().isDocumentUnsaved(doc)) {
      msg.append("[+] ");
    }

    int lline = editor.getCaretModel().getLogicalPosition().line;
    int total = EditorHelper.getLineCount(editor);
    int pct = (int)((float)lline / (float)total * 100f + 0.5);

    msg.append("line ").append(lline + 1).append(" of ").append(total);
    msg.append(" --").append(pct).append("%-- ");

    LogicalPosition lp = editor.getCaretModel().getLogicalPosition();
    int col = editor.getCaretModel().getOffset() - doc.getLineStartOffset(lline);

    msg.append("col ").append(col + 1);
    if (col != lp.column) {
      msg.append("-").append(lp.column + 1);
    }

    VimPlugin.showMessage(msg.toString());
  }

  @NotNull private static final String disposableKey = "VimFileGroupDisposable";

  @NotNull private static final HashMap<FileEditorManager, VirtualFile> lastSelections = new HashMap<>();
  @NotNull private static final Logger logger = Logger.getInstance(FileGroup.class.getName());

  /**
   * This method listens for editor tab changes so any insert/replace modes that need to be reset can be.
   */
  public static void fileEditorManagerSelectionChangedCallback(@NotNull FileEditorManagerEvent event) {
    // The user has changed the editor they are working with - exit insert/replace mode, and complete any
    // appropriate repeat
    if (event.getOldFile() != null) {
      lastSelections.put(event.getManager(), event.getOldFile());
      String disposableKey = FileGroup.disposableKey + event.getManager().hashCode();
      if (Disposer.get(disposableKey) == null) {
        Disposer.register(event.getManager().getProject(), () -> {
          lastSelections.remove(event.getManager());
        }, disposableKey);
      }
    }
  }
}

File: src/main/java/com/maddyhome/idea/vim/group/FileGroup.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.group

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.FileEditorManagerEvent
import com.intellij.openapi.fileEditor.TextEditor
import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx
import com.intellij.openapi.fileEditor.impl.EditorsSplitters
import com.intellij.openapi.fileTypes.FileTypeManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.roots.ProjectRootManager
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.psi.search.FilenameIndex
import com.intellij.psi.search.ProjectScope
import com.maddyhome.idea.vim.VimPlugin
import com.maddyhome.idea.vim.api.ExecutionContext
import com.maddyhome.idea.vim.api.VimEditor
import com.maddyhome.idea.vim.api.VimFileBase
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.common.TextRange
import com.maddyhome.idea.vim.group.LastTabService.Companion.getInstance
import com.maddyhome.idea.vim.helper.EditorHelper
import com.maddyhome.idea.vim.helper.MessageHelper.message
import com.maddyhome.idea.vim.helper.countWords
import com.maddyhome.idea.vim.helper.fileSize
import com.maddyhome.idea.vim.newapi.IjEditorExecutionContext
import com.maddyhome.idea.vim.newapi.IjVimEditor
import com.maddyhome.idea.vim.newapi.execute
import com.maddyhome.idea.vim.newapi.globalIjOptions
import com.maddyhome.idea.vim.state.VimStateMachine.Companion.getInstance
import com.maddyhome.idea.vim.state.mode.Mode.VISUAL
import java.io.File
import java.util.*

class FileGroup : VimFileBase() {
  override fun openFile(filename: String, context: ExecutionContext): Boolean {
    if (logger.isDebugEnabled) {
      logger.debug("openFile($filename)")
    }
    val project = PlatformDataKeys.PROJECT.getData((context as IjEditorExecutionContext).context)
      ?: return false // API change - don't merge

    val found = findFile(filename, project)

    if (found != null) {
      if (logger.isDebugEnabled) {
        logger.debug("found file: $found")
      }
      // Can't open a file unless it has a known file type. The next call will return the known type.
      // If unknown, IDEA will prompt the user to pick a type.
      val type = FileTypeManager.getInstance().getKnownFileTypeOrAssociate(found, project)

      if (type != null) {
        val fem = FileEditorManager.getInstance(project)
        fem.openFile(found, true)

        return true
      } else {
        // There was no type and user didn't pick one. Don't open the file
        // Return true here because we found the file but the user canceled by not picking a type.
        return true
      }
    } else {
      VimPlugin.showMessage(message("unable.to.find.0", filename))

      return false
    }
  }

  fun findFile(filename: String, project: Project): VirtualFile? {
    var found: VirtualFile?
    // Vim supports both ~/ and ~\ (tested on Mac and Windows). On Windows, it supports forward- and back-slashes, but
    // it only supports forward slash on Unix (tested on Mac)
    // VFS works with both directory separators (tested on Mac and Windows)
    if (filename.startsWith("~/") || filename.startsWith("~\\")) {
      val relativePath = filename.substring(2)
      val dir = System.getProperty("user.home")
      if (logger.isDebugEnabled) {
        logger.debug("home dir file")
        logger.debug("looking for $relativePath in $dir")
      }
      found = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(File(dir, relativePath))
    } else {
      found = LocalFileSystem.getInstance().findFileByIoFile(File(filename))

      if (found == null) {
        found = findByNameInContentRoots(filename, project)
        if (found == null) {
          found = findByNameInProject(filename, project)
        }
      }
    }

    return found
  }

  private fun findByNameInContentRoots(filename: String, project: Project): VirtualFile? {
    var found: VirtualFile? = null
    val prm = ProjectRootManager.getInstance(project)
    val roots = prm.contentRoots
    for (i in roots.indices) {
      if (logger.isDebugEnabled) {
        logger.debug("root[" + i + "] = " + roots[i].path)
      }
      found = roots[i].findFileByRelativePath(filename)
      if (found != null) {
        break
      }
    }
    return found
  }

  /**
   * Closes the current editor.
   */
  override fun closeFile(editor: VimEditor, context: ExecutionContext) {
    val project = PlatformDataKeys.PROJECT.getData((context.context as DataContext))
    if (project != null) {
      val fileEditorManager = FileEditorManagerEx.getInstanceEx(project)
      val window = fileEditorManager.currentWindow
      val virtualFile = fileEditorManager.currentFile

      if (virtualFile != null && window != null) {
        // During the work on VIM-2912 I've changed the close function to this one.
        //   However, the function with manager seems to work weirdly and it causes VIM-2953
        //window.getManager().closeFile(virtualFile, true, false);
        window.closeFile(virtualFile)

        // Get focus after closing tab
        window.requestFocus(true)
        if (!ApplicationManager.getApplication().isUnitTestMode) {
          // This thing doesn't have an implementation in test mode
          EditorsSplitters.focusDefaultComponentInSplittersIfPresent(project)
        }
      }
    }
  }

  /**
   * Closes editor.
   */
  override fun closeFile(number: Int, context: ExecutionContext) {
    val project = PlatformDataKeys.PROJECT.getData((context as IjEditorExecutionContext).context) ?: return
    val fileEditorManager = FileEditorManagerEx.getInstanceEx(project)
    val window = fileEditorManager.currentWindow
    val editors = fileEditorManager.openFiles
    if (window != null) {
      if (number >= 0 && number < editors.size) {
        fileEditorManager.closeFile(editors[number], window)
      }
    }
    if (!ApplicationManager.getApplication().isUnitTestMode) {
      // This thing doesn't have an implementation in test mode
      EditorsSplitters.focusDefaultComponentInSplittersIfPresent(project)
    }
  }

  /**
   * Saves specific file in the project.
   */
  override fun saveFile(context: ExecutionContext) {
    val action = if (injector.globalIjOptions().ideawrite.contains(IjOptionConstants.ideawrite_all)) {
      injector.nativeActionManager.saveAll
    } else {
      injector.nativeActionManager.saveCurrent
    }
    action.execute(context)
  }

  /**
   * Saves all files in the project.
   */
  override fun saveFiles(context: ExecutionContext) {
    injector.nativeActionManager.saveAll.execute(context)
  }

  /**
   * Selects then next or previous editor.
   */
  override fun selectFile(count: Int, context: ExecutionContext): Boolean {
    var count = count
    val project = PlatformDataKeys.PROJECT.getData((context as IjEditorExecutionContext).context) ?: return false
    val fem = FileEditorManager.getInstance(project) // API change - don't merge
    val editors = fem.openFiles
    if (count == 99) {
      count = editors.size - 1
    }
    if (count < 0 || count >= editors.size) {
      return false
    }

    fem.openFile(editors[count], true)

    return true
  }

  /**
   * Selects then next or previous editor.
   */
  override fun selectNextFile(count: Int, context: ExecutionContext) {
    val project = PlatformDataKeys.PROJECT.getData((context as IjEditorExecutionContext).context) ?: return
    val fem = FileEditorManager.getInstance(project) // API change - don't merge
    val editors = fem.openFiles
    val current = fem.selectedFiles[0]
    for (i in editors.indices) {
      if (editors[i] == current) {
        val pos = (i + (count % editors.size) + editors.size) % editors.size

        fem.openFile(editors[pos], true)
      }
    }
  }

  /**
   * Selects previous editor tab.
   */
  override fun selectPreviousTab(context: ExecutionContext) {
    val project = PlatformDataKeys.PROJECT.getData((context.context as DataContext)) ?: return
    val vf = getInstance(project).lastTab
    if (vf != null && vf.isValid) {
      FileEditorManager.getInstance(project).openFile(vf, true)
    } else {
      VimPlugin.indicateError()
    }
  }

  /**
   * Returns the previous tab.
   */
  fun getPreviousTab(context: DataContext): VirtualFile? {
    val project = PlatformDataKeys.PROJECT.getData(context) ?: return null
    val vf = getInstance(project).lastTab
    if (vf != null && vf.isValid) {
      return vf
    }
    return null
  }

  fun selectEditor(project: Project, file: VirtualFile): Editor? {
    val fMgr = FileEditorManager.getInstance(project)
    val feditors = fMgr.openFile(file, true)
    if (feditors.size > 0) {
      if (feditors[0] is TextEditor) {
        val editor = (feditors[0] as TextEditor).editor
        if (!editor.isDisposed) {
          return editor
        }
      }
    }

    return null
  }

  override fun displayLocationInfo(vimEditor: VimEditor) {
    val editor = (vimEditor as IjVimEditor).editor
    val msg = StringBuilder()
    val doc = editor.document

    if (injector.vimState.mode !is VISUAL) {
      val lp = editor.caretModel.logicalPosition
      val col = editor.caretModel.offset - doc.getLineStartOffset(lp.line)
      var endoff = doc.getLineEndOffset(lp.line)
      if (endoff < editor.fileSize && doc.charsSequence[endoff] == '\n') {
        endoff--
      }
      val ecol = endoff - doc.getLineStartOffset(lp.line)
      val elp = editor.offsetToLogicalPosition(endoff)

      msg.append("Col ").append(col + 1)
      if (col != lp.column) {
        msg.append("-").append(lp.column + 1)
      }

      msg.append(" of ").append(ecol + 1)
      if (ecol != elp.column) {
        msg.append("-").append(elp.column + 1)
      }

      val lline = editor.caretModel.logicalPosition.line
      val total = IjVimEditor(editor).lineCount()

      msg.append("; Line ").append(lline + 1).append(" of ").append(total)

      val cp = countWords(vimEditor)

      msg.append("; Word ").append(cp.position).append(" of ").append(cp.count)

      val offset = editor.caretModel.offset
      val size = editor.fileSize

      msg.append("; Character ").append(offset + 1).append(" of ").append(size)
    } else {
      msg.append("Selected ")

      val vr = TextRange(
        editor.selectionModel.blockSelectionStarts,
        editor.selectionModel.blockSelectionEnds
      )
      vr.normalize()

      val lines: Int
      var cp = countWords(vimEditor)
      val words = cp.count
      var word = 0
      if (vr.isMultiple) {
        lines = vr.size()
        val cols = vr.maxLength

        msg.append(cols).append(" Cols; ")

        for (i in 0 until vr.size()) {
          cp = countWords(vimEditor, vr.startOffsets[i], (vr.endOffsets[i] - 1).toLong())
          word += cp.count
        }
      } else {
        val slp = editor.offsetToLogicalPosition(vr.startOffset)
        val elp = editor.offsetToLogicalPosition(vr.endOffset)

        lines = elp.line - slp.line + 1

        cp = countWords(vimEditor, vr.startOffset, (vr.endOffset - 1).toLong())
        word = cp.count
      }

      val total = IjVimEditor(editor).lineCount()

      msg.append(lines).append(" of ").append(total).append(" Lines")

      msg.append("; ").append(word).append(" of ").append(words).append(" Words")

      val chars = vr.selectionCount
      val size = editor.fileSize

      msg.append("; ").append(chars).append(" of ").append(size).append(" Characters")
    }

    VimPlugin.showMessage(msg.toString())
  }

  override fun displayFileInfo(vimEditor: VimEditor, fullPath: Boolean) {
    val editor = (vimEditor as IjVimEditor).editor
    val msg = StringBuilder()
    val vf = EditorHelper.getVirtualFile(editor)
    if (vf != null) {
      msg.append('"')
      if (fullPath) {
        msg.append(vf.path)
      } else {
        val project = editor.project
        if (project != null) {
          val root = ProjectRootManager.getInstance(project).fileIndex.getContentRootForFile(vf)
          if (root != null) {
            msg.append(vf.path.substring(root.path.length + 1))
          } else {
            msg.append(vf.path)
          }
        }
      }
      msg.append("\" ")
    } else {
      msg.append("\"[No File]\" ")
    }

    val doc = editor.document
    if (!doc.isWritable) {
      msg.append("[RO] ")
    } else if (FileDocumentManager.getInstance().isDocumentUnsaved(doc)) {
      msg.append("[+] ")
    }

    val lline = editor.caretModel.logicalPosition.line
    val total = IjVimEditor(editor).lineCount()
    val pct = (lline.toFloat() / total.toFloat() * 100f + 0.5).toInt()

    msg.append("line ").append(lline + 1).append(" of ").append(total)
    msg.append(" --").append(pct).append("%-- ")

    val lp = editor.caretModel.logicalPosition
    val col = editor.caretModel.offset - doc.getLineStartOffset(lline)

    msg.append("col ").append(col + 1)
    if (col != lp.column) {
      msg.append("-").append(lp.column + 1)
    }

    VimPlugin.showMessage(msg.toString())
  }

  override fun selectEditor(projectId: String, documentPath: String, protocol: String?): VimEditor? {
    val fileSystem = VirtualFileManager.getInstance().getFileSystem(protocol) ?: return null
    val virtualFile = fileSystem.findFileByPath(documentPath) ?: return null

    val project = Arrays.stream(ProjectManager.getInstance().openProjects)
      .filter { p: Project? -> injector.file.getProjectId(p!!) == projectId }
      .findFirst().orElseThrow()

    val editor = selectEditor(project, virtualFile) ?: return null
    return IjVimEditor(editor)
  }

  override fun getProjectId(project: Any): String {
    require(project is Project)
    return project.name + "-" + project.locationHash
  }

  companion object {
    private fun findByNameInProject(filename: String, project: Project): VirtualFile? {
      val projectScope = ProjectScope.getProjectScope(project)
      val names = FilenameIndex.getVirtualFilesByName(filename, projectScope)
      if (!names.isEmpty()) {
        return names.stream().findFirst().get()
      }
      return null
    }

    private val logger = Logger.getInstance(
      FileGroup::class.java.name
    )

    /**
     * Respond to editor tab selection and remember the last used tab
     */
    fun fileEditorManagerSelectionChangedCallback(event: FileEditorManagerEvent) {
      if (event.oldFile != null) {
        getInstance(event.manager.project).lastTab = event.oldFile
      }
    }
  }
}
